crates/astria-cli/src/commands/bridge.rs

```diff
@@ -0,0 +1,484 @@
+    SequencerWithdrawalFilter,
+};
+use astria_core::{
+    bridge::{
+        self,
+        Ics20WithdrawalFromRollupMemo,
+    },
+    primitive::v1::{
+        asset::{
+            self,
+            TracePrefixed,
+        },
+        Address,
+    },
+    protocol::transaction::v1alpha1::{
+        action::{
+            BridgeUnlockAction,
+            Ics20Withdrawal,
+        },
+        Action,
+    },
+};
+use clap::Args;
+use color_eyre::eyre::{
+    self,
+    bail,
+    eyre,
+    OptionExt as _,
+    WrapErr as _,
+};
+use ethers::{
+    contract::EthEvent,
+    core::types::Block,
+    providers::{
+        Middleware,
+        Provider,
+        ProviderError,
+        StreamExt as _,
+        Ws,
+    },
+    types::{
+        Filter,
+        Log,
+        H256,
+    },
+};
+use tokio::select;
+use tracing::warn;
+
+#[derive(Args, Debug)]
+pub struct CollectWithdrawalEvents {
+    #[arg(long)]
+    rollup_endpoint: String,
+    #[arg(long)]
+    contract_address: ethers::types::Address,
+    #[arg(long)]
+    next_rollup_block_height: u64,
+    #[arg(long, default_value = "nria")]
+    fee_asset: asset::Denom,
+    #[arg(long)]
+    rollup_asset_denom: asset::Denom,
+    #[arg(long)]
+    bridge_address: Address,
+    #[arg(long, default_value = "astria")]
+    sequencer_address_prefix: String,
+    #[arg(long, short)]
+    output: PathBuf,
+}
+
+impl CollectWithdrawalEvents {
+    pub(crate) async fn run(self) -> eyre::Result<()> {
+        let Self {
+            rollup_endpoint,
+            contract_address,
+            next_rollup_block_height,
+            fee_asset,
+            rollup_asset_denom,
+            bridge_address,
+            sequencer_address_prefix,
+            output,
+        } = self;
+
+        let output_file = std::fs::File::options()
+            .write(true)
+            .create_new(true)
+            .open(&output)
+            .wrap_err("failed to open specified file for writing")?;
+
+        let block_provider = connect_to_rollup(&rollup_endpoint)
+            .await
+            .wrap_err("failed to connect to rollup")?;
+
+        let asset_withdrawal_divisor =
+            get_asset_withdrawal_divisor(contract_address, block_provider.clone())
+                .await
+                .wrap_err("failed determining asset withdrawal divisor")?;
+
+        let mut block_subscription = block_provider
+            .subscribe_blocks()
+            .await
+            .wrap_err("failed to subscribe to blocks from rollup")?
+            .boxed();
+
+        let Some(current_rollup_block) = block_subscription.next().await else {
+            bail!("failed to get current rollup block from subscription")
+        };
+
+        let Some(current_rollup_block_height) = current_rollup_block.number else {
+            bail!(
+                "couldn't determine current rollup block height; value was not set on current on \
+                 most recent block",
+            );
+        };
```

Comment:
you don't use the subscription anywhere right? you can just get the latest block and use the number from that instead: https://docs.rs/ethers/latest/ethers/middleware/trait.Middleware.html#method.get_block

Reply:
The subscription gets used further down in the chained stream (not correct code, just for demonstration purposes):
```rust
let incoming_blocks = iter(start_height..current_hight).then(|height| providers.get_bock(height)),chain(block_subscription);
```
Doing it this way is import to not have the rollup advance between getting `current_height`, fetching blocks from `start` to `current_height`, and then going back to the stream.

From a first glance the ethers subscription buffers blocks so that we are not going to miss out